minor cleanups and rewriting middleware documentation with examples (#912)
strapi/documentation: docs/developer-docs/latest/setup-deployment-guides/configurations/required/middlewares.md

module.exports = [
  // The array is pre-populated with internal, built-in middlewares, prefixed by `strapi::`
  'strapi::errors',
  'strapi::security',
  'strapi::cors',

  // custom middleware that does not require any configuration
  'my-custom-node-module', 

  // custom name to find a package or a path
  {
    name: 'my-custom-node-module',
    config: {
      foo: 'bar',
    },
  },

  // custom resolve to find a package or a path
  {
    resolve: '../some-dir/custom-middleware',
    config: {
      foo: 'bar',
    },
  },

>   // custom config for internal & built-in middleware

Author: We use both in this document, I see no difference myself

- and [session](#session).

::: caution

> The following built-in middlewares are automatically added by Strapi: `errors`, `security`, `cors`, `query`, `body`, `public`, `session`, `favicon`. They should not be removed as it will throw an error.

Reviewer: The session middleware is added by default but is not a required middleware

Reviewer 2: I'll rephrase, for clarity.

> The following built-in middlewares are automatically added by Strapi: `errors`, `security`, `cors`, `query`, `body`, `public`, `session`, `favicon`. They should not be removed as it will throw an error.

Reviewer 2: ```suggestion
The following built-in middlewares are automatically added by Strapi: `errors`, `security`, `cors`, `query`, `body`, `public`, `session`, `favicon`. All of them except `session` should not be removed as it will throw an error.
```

- and [session](#session).

::: caution
The following built-in middlewares are automatically added by Strapi: `errors`, `security`, `cors`, `query`, `body`, `public`, `session`, `favicon`. They should not be removed as it will throw an error.
:::

### `body`

The `body` middleware is based on [koa-body](https://github.com/koajs/koa-body). It accepts the following options:

| Option       | Description                                        | Type                  | Default     |
|--------------|----------------------------------------------------|-----------------------|-------------|
| `multipart`  | Parse multipart bodies                             | `Boolean`             | `true`      |
| `patchKoa`   | Patch request body to Koa's `ctx.request`          | `Boolean`             | `true`      |

> | `jsonLimit`  | The byte (if integer) limit of the JSON body       | `String` or `Integer` | `1mb`       |

Author: You can either pass in a byte value (integer) or a string with a different incremental connotation.

Eg: `1mb` === `1000000` === `1000kb` they are all the same

| `hsts`                      | Set options for the HTTP Strict Transport Security (HSTS) policy.                             | `Object`              | See below     |
| `hsts.maxAge`               | Number of seconds HSTS is in effect                                                           | `Integer`             | `31536000`    |
| `hsts.includeSubDomains`    | Applies HSTS to all subdomains of the host                                                    | `Boolean`             | `true`        |
| `frameguard`                | Set `X-Frame-Options` header to help mitigate clickjacking attacks, set to false to disable   | `Boolean` or `Object` | See below     |
| `frameguard.action`         | Value must be either `deny` or `sameorigin`                                                   | `String`              | `sameorigin`  |

::: tip
When using any 3rd party upload provider, generally it's required to set a custom configuration here. Please see the provider documentation for which configuration options are required.
:::

> ::: strapi

Reviewer 2: ```suggestion
::: note
```

This should be a note callout :-) `strapi` callouts are for very specific use cases, such as highlighting some important concepts or explaining differences between two misunderstood topics, and then cross-linking to other sections of the documentation.

> Within the default directives there is a `dl.airtable.com`, this value is set for the [in-app market](/user-docs/latest/plugins/installing-plugins-via-marketplace.md) and is absolutely safe to keep.

Reviewer 2: ```suggestion
The default directives include a `dl.airtable.com` value. This value is set for the [in-app market](/user-docs/latest/plugins/installing-plugins-via-marketplace.md) and is safe to keep.
```

            'data:',
            'blob:',
            'dl.airtable.com',
            'yourBucketName.s3.yourRegion.amazonaws.com',
          ],
          'media-src': [
            "'self'",
            'data:',
            'blob:',
            'dl.airtable.com',
            'yourBucketName.s3.yourRegion.amazonaws.com',
          ],
          upgradeInsecureRequests: null,
        },
      },
    },
  },
  // ...
}
```

:::

### `session`

The `session` middleware allows the use of cookie-based sessions, based on [koa-session](https://github.com/koajs/session). It accepts the following options:

| Option       | Description                                                                                                  | Type                     | Default value                           |
|--------------|--------------------------------------------------------------------------------------------------------------|--------------------------|-----------------------------------------|
| `key`        | Cookie key                                                                                                   | `String`                 | `'koa.sess'`                            |

> | `maxAge`     | Maximum lifetime of the cookies, in ms. Using `'session'` will expire the cookie when the session is closed. | `Integer` or `'session'` | `86400000`                              |

Reviewer 2: ```suggestion
| `maxAge`     | Maximum lifetime of the cookies, in milliseconds. Using `'session'` will expire the cookie when the session is closed. | `Integer` or `'session'` | `86400000`                              |
```
